packages/test/test-end-to-end-tests/src/test/audience.spec.ts

```diff
@@ -0,0 +1,156 @@
+            this.audienceClientList.add(clientId);
+        }
+    }
+}
+
+describeNoCompat("Audience correctness", (getTestObjectProvider) => {
+    let provider: ITestObjectProvider;
+    const dataObjectFactory = new DataObjectFactory(
+        "TestDataObject",
+        TestDataObject,
+        [],
+        []);
+    const runtimeFactory = new ContainerRuntimeFactoryWithDefaultDataStore(
+        dataObjectFactory,
+        [
+            [dataObjectFactory.type, Promise.resolve(dataObjectFactory)],
+        ],
+        undefined,
+        undefined,
+        // Disable summaries so the summarizer client doesn't interfere with the audience
+        flattenRuntimeOptions({ summaryOptions: { generateSummaries: false }}),
+    );
+
+    const createContainer = async (): Promise<Container> => await provider.createContainer(runtimeFactory) as Container;
+    const loadContainer = async (): Promise<Container> => await provider.loadContainer(runtimeFactory) as Container;
+
+    beforeEach(async () => {
+        provider = getTestObjectProvider();
+    });
+
+    it("should add / remove clients in audience as expected", async () => {
+        // Create a client - client1.
+        const client1Container = await createContainer();
+        const client1DataStore = await requestFluidObject<TestDataObject>(client1Container, "default");
+
+        // Ensure the client1 is connected and synchronized.
+        await provider.ensureSynchronized();
+        // Validate that client1 is in its own audience.
+        assert(client1Container.clientId !== undefined, "client1 does not have clientId");
+        assert(
+            client1DataStore.audienceClientList.has(client1Container.clientId),
+            "client1's audience does not have client1's clientId",
+        );
+
+        // Load a second client - client2.
+        const client2Container = await loadContainer();
+        const client2DataStore = await requestFluidObject<TestDataObject>(client2Container, "default");
+        // Ensure the client2 is connected and synchronized.
+        await provider.ensureSynchronized();
+        // Validate that client2 is in its own audience.
+        assert(client2Container.clientId !== undefined, "client does not have clientId");
+        assert(
+            client2DataStore.audienceClientList.has(client2Container.clientId),
+            "client2's audience does not have client2's clientId",
+        );
+
+        // Validate that client1 is in client2's audience.
+        assert(
+            client2DataStore.audienceClientList.has(client1Container.clientId),
+            "client2's audience does not have client1",
+        );
+
+        // Validate that client2 is in client1's audience.
+        assert(
+            client1DataStore.audienceClientList.has(client2Container.clientId),
+            "Client1's audience does not have client2",
+        );
+    });
+
+    it("should add / remove clients in audience as expected in write mode", async () => {
```

Comment:
We are not removing any audience member in this test, so maybe fix the test name or you can have another test to test removal of audience memeber?

Reply:
Good point. Acutally, transition from read -> write mode does a remove followed by add. But yeah, I should add a separate test where I verify the removed client is gone from audience.

Reply:
Also maybe we want to add an assert to make sure we were in read mode previously as then only we are testing the transition.

Reply:
That's not super important. As long as we ensure the client is in write mode, it should be sufficient.